verification.md

```diff
@@ -11,24 +12,296 @@ module VERIFICATION
+
+Instead, we abstract it as an uninterpreted function that captures the possibility of the hash collision, even if the probability is very small.
+
+```
+  syntax Int ::= keccak ( WordStack ) [function, smtlib(smt_keccak)]
+```
+
+However, one can avoid reasoning about the potential collision by assuming all of the hashed values appearing in each execution trace are collision-free. This can be achieved by instantiating the injectivity property only for the terms appearing in the symbolic execution, in a way analogous to the universal quantifier instantiation.
+
+#### Hashed Locations
+
+Hashed locations are essential for the storage layout schemes used to store compound data structures such as maps in the storage. For more details, refer to [EVM-Specific EVM Specification](proofs/erc20.md#storage).
+
+The following syntactic sugars capture the storage layout schemes of Solidity and Viper.
+
+```{.k .uiuck}
+  syntax IntList ::= List{Int, ""} [klabel(intList)]
+
+  syntax Int ::= #hashedLocation(String, Int, IntList) [function]
+  rule #hashedLocation(LANG, BASE, .IntList) => BASE
+  //
+  rule #hashedLocation("Viper",    BASE, OFFSET OFFSETS) => #hashedLocation("Viper",    keccakIntList(BASE) +Word OFFSET, OFFSETS)
+  rule #hashedLocation("Solidity", BASE, OFFSET OFFSETS) => #hashedLocation("Solidity", keccakIntList(OFFSET BASE),       OFFSETS)
+
+  syntax Int ::= keccakIntList(IntList) [function]
+  rule keccakIntList(VS) => keccak(intList2ByteStack(VS))
+
+  syntax WordStack ::= intList2ByteStack(IntList) [function]
+  rule intList2ByteStack(.IntList) => .WordStack
+  rule intList2ByteStack(V VS) => #asByteStackInWidth(V, 32) ++ intList2ByteStack(VS)
+    requires 0 <=Int V andBool V <Int (2 ^Int 256)
+```
+
+### Arithmetic Simplification Rules
+
+We introduce simplification rules that capture arithmetic properties, which reduce the given terms into smaller ones. These rules help to improve the performance of the underlying theorem prover’s symbolic reasoning.
+
+Below are universal simplification rules that are free to be used in any context.
+
+```{.k .uiuck}
+  rule 0 +Int N => N
+  rule N +Int 0 => N
+
+  rule N -Int 0 => N
+
+  rule 1 *Int N => N
+  rule N *Int 1 => N
+  rule 0 *Int _ => 0
+  rule _ *Int 0 => 0
+
+  rule N /Int 1 => N
+
+  rule 0 |Int N => N
+  rule N |Int 0 => N
+  rule N |Int N => N
+
+  rule 0 &Int N => 0
+  rule N &Int 0 => 0
+  rule N &Int N => N
+```
+
+The following simplification rules are local, meant to be used in specific contexts. The rules are applied only when the side-conditions are met.
+
+These rules are specific to reasoning about EVM programs.
 
 ```{.k .uiuck}
-    rule #sizeWordStack ( _ , _ ) >=Int 0 => true [smt-lemma]
-    rule #sizeWordStack ( WS , N:Int )
-      => #sizeWordStack ( WS , 0 ) +Int N
-      requires N =/=K 0
-      [lemma]
+  syntax Bool ::= #isConcrete(K) [function, hook(KREFLECTION.isConcrete)]
```

Comment:
hm, so what's the relationship between this and the the `[concrete]` attribute?

Reply:
`[concrete]` attribute is a syntactic sugar for `#isConcrete`.